Jellyfin.Server.Implementations/MediaSegments/MediaSegmentManager.cs

```diff
@@ -0,0 +1,104 @@
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading.Tasks;
+using Jellyfin.Data.Entities;
+using Jellyfin.Data.Enums;
+using MediaBrowser.Controller;
+using MediaBrowser.Controller.Entities;
+using MediaBrowser.Model.MediaSegments;
+using Microsoft.EntityFrameworkCore;
+
+namespace Jellyfin.Server.Implementations;
+
+/// <summary>
+///     Manages media segments retrival and storage.
+/// </summary>
+public class MediaSegmentManager : IMediaSegmentManager
+{
+    private readonly IDbContextFactory<JellyfinDbContext> _dbProvider;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="MediaSegmentManager"/> class.
+    /// </summary>
+    /// <param name="dbProvider">EFCore Database factory.</param>
+    public MediaSegmentManager(IDbContextFactory<JellyfinDbContext> dbProvider)
+    {
+        _dbProvider = dbProvider;
+    }
+
+    /// <inheritdoc />
+    public async Task<MediaSegmentDto> CreateSegmentAsync(MediaSegmentDto mediaSegment)
+    {
+        ArgumentOutOfRangeException.ThrowIfLessThan(mediaSegment.EndTicks, mediaSegment.StartTicks);
+
+        using var db = await _dbProvider.CreateDbContextAsync().ConfigureAwait(false);
+        db.MediaSegments.Add(Map(mediaSegment));
+        await db.SaveChangesAsync().ConfigureAwait(false);
+        return mediaSegment;
+    }
+
+    /// <inheritdoc />
+    public async Task DeleteSegmentAsync(Guid segmentId)
+    {
+        using var db = await _dbProvider.CreateDbContextAsync().ConfigureAwait(false);
+        await db.MediaSegments.Where(e => e.Id.Equals(segmentId)).ExecuteDeleteAsync().ConfigureAwait(false);
+    }
+
+    /// <inheritdoc />
+    public async Task<IEnumerable<MediaSegmentDto>> GetSegmentsAsync(Guid itemId, IEnumerable<MediaSegmentType>? typeFilter)
+    {
+        using var db = await _dbProvider.CreateDbContextAsync().ConfigureAwait(false);
+
+        var query = db.MediaSegments
+            .Where(e => e.ItemId.Equals(itemId));
+
+        if (typeFilter is not null)
+        {
+            query = query.Where(e => typeFilter.Contains(e.Type));
+        }
+
+        return query
+            .OrderBy(e => e.StartTicks)
+            .ToImmutableList()
```

Comment:
Why `ToImmutableList`?

Reply:
List because the original size of the source is unknown and there is a small performance and memory advantage if you use the Immutable variant if there is no need for modifications which isn't here as the result will be discarded as soon as the mapping happens